Allow authors to remove option for unproctored exams

andy-armstrong opened the pull request:
https://openedx.atlassian.net/browse/TNL-5084

This work is in support of the new edx-platform feature here: https://github.com/edx/edx-platform/pull/13840

@dianakhuang @robrap please review.

robrap commented on edx_proctoring/api.py:
```diff
@@ -1769,6 +1768,7 @@ def _get_proctored_exam_view(exam, context, exam_id, user_id, course_id):
         student_view_template = 'proctored_exam/ready_to_submit.html'
 
     if student_view_template:
+        log.info("Proctoring context: %s", context)
```
Same logging question as on [the other PR](https://github.com/edx/edx-platform/pull/13840#discussion_r85556862).  If no change is required there, none here either.  We'll see where that goes.

andy-armstrong replied:
I added both of these to debug a sandbox issue. I think it is too verbose, so I'm removing them now.

robrap commented on edx_proctoring/templates/proctored_exam/entrance.html:
```diff
@@ -20,15 +20,18 @@
   </button>
-  <button class="gated-sequence start-timed-exam" data-attempt-proctored=false>
+
+  {% if allow_unproctored %}
```
Same comment around `unproctored`.  I'd recommend doing a search and fixing all instances once you have an and appropriate name (e.g. `allow_proctoring_opt_out`).

robrap commented on edx_proctoring/templates/proctored_exam/failed-prerequisites.html:
```diff
@@ -26,23 +26,26 @@
+
+  {% if allow_unproctored %}
+    <p>
     {% blocktrans %}
```
Indentation is inconsistent with code below.

robrap commented on edx_proctoring/templates/proctored_exam/failed-prerequisites.html:
```diff
@@ -26,23 +26,26 @@
+    </p>
 
-  <button class="gated-sequence start-timed-exam" data-attempt-proctored=false>
+    <button class="gated-sequence start-timed-exam" data-attempt-proctored=false>
```
It feels like this should be pulled out as a separate included template.  The redundancy leads to issues like the language needlessly differing between: `I understand that I am not eligible for academic credit.` and `I am not interested in academic credit.`.

I could imagine the `Due to unsatisfied prerequisites...` comment remaining here in a separate if, or moving into the shared template with a conditional.

You may not like this idea, so feel free to discuss before diving in.

andy-armstrong replied:
@robrap thanks for the suggestion. I've made a shared template and used it in three of the four places. Let me know if it is what you had in mind.

robrap commented on edx_proctoring/templates/proctored_exam/instructions.html:
```diff
@@ -79,12 +79,12 @@ <h2 id="acessible-error-title">{% trans "Cannot Start Proctored Exam" %}</h2>
+  {% if not is_sample_attempt and allow_unproctored %}
+    <p class="proctored-exam-instruction">
       {% trans "Don't want to take this exam with online proctoring?" %}<a class="proctored-decline-exam" href='#' data-action="decline" data-exam-id="{{exam_id}}" data-change-state-url="{{change_state_url}}">
         {% trans "Take this exam as an open exam instead." %}
```
I'm confused about the purpose of this?  Why wouldn't this want similar text?  Is an "open exam" different from an "exam without proctoring"?  Wouldn't this want the same warning about academic credit?  I know this wasn't added by your change, but I think discussing a little clean-up is warranted, even if we decide to do nothing.

catong replied:
I can't run through the workflow on the sandbox, so I can't verify which templates are used. Consolidation/clean up would be great; it seems that we might have either outdated or unused templates? Wording does not match the screenshot of the template that's currently in the Learner documentation: http://edx.readthedocs.io/projects/edx-guide-for-students/en/latest/completing_assignments/SFD_proctored_exams.html#agree-to-take-the-proctored-exam

andy-armstrong replied:
@catong I discussed this with @dianakhuang and our conclusion is that these represent two distinct use cases. The other button is shown when you haven't yet chosen whether to choose proctored or 'unproctored':

![image](https://cloud.githubusercontent.com/assets/5985072/19821876/b412e7b6-9d2e-11e6-9791-3d5a76a24a2e.png)

Once you have chosen proctored, you are presented at the bottom of the instructions with the link we are discussing here to change your mind:

![image](https://cloud.githubusercontent.com/assets/5985072/19821787/479864c6-9d2e-11e6-9625-20859c607b23.png)

The UI is quite different so it would be hard to share the logic. However, I'm happy to tweak the wording to make the language more consistent as you see fit.

robrap replied:
Thanks for the screen shots.  This really helps.  I'd highly recommend not introducing a different term like "open exam", as well as including the warning about academic credit.

andy-armstrong replied:
@catong can you recommend a revised version of this footer. Thanks!

catong replied:
@andy-armstrong @robrap How about:
"Take this exam without proctoring. Doing so means that you are no longer eligible for academic credit." 
where "Take this exam without proctoring" is the link text.

robrap commented on edx_proctoring/tests/test_student_view.py:
```diff
@@ -55,98 +55,121 @@ def setUp(self):
         self.chose_proctored_exam_msg = 'Follow these steps to set up and start your proctored exam'
         self.proctored_exam_optout_msg = 'Take this exam as an open exam instead'
         self.proctored_exam_completed_msg = 'Are you sure you want to end your proctored exam'
+        self.proctored_exam_waiting_for_app_shutdown_msg = 'You are about to complete your proctored exam'
```
This line should be gone.  Maybe a resolve conflict gone bad.

andy-armstrong replied:
Whoops, my bad. I've removed it.

robrap commented on edx_proctoring/tests/test_student_view.py:
```diff
@@ -161,29 +184,47 @@ def test_verification_status(self, verification_status, expected_message):
-        )
+    def test_proctored_only_entrance(self):
+        """
+        This test verifies that learners are not given the option to enter an unproctored exam
```
For comments, even thought it is longer, I think referring to a "proctored exam without proctoring" is more accurate than an "unproctored" exam, which I would interpret as a "timed exam".

andy-armstrong replied:
Agreed. I've made this change in all the doc strings.

robrap commented on edx_proctoring/tests/test_student_view.py:
```diff
@@ -55,98 +55,121 @@ def setUp(self):
+            context=context,
+        )
+
+    def render_proctored_exam(self, context_overrides=None):
```
Thank you for all the code cleanup!

robrap commented on edx_proctoring/tests/test_student_view.py:
```diff
@@ -365,62 +407,36 @@ def test_get_studentview_unstarted_exam(self):
 
-        # now make sure content remains the same if
-        # the status transitions to 'download_software_clicked'
+        # Verify that the option to skip proctoring is not presented
```
How do you decide when to add a new test vs adding the opposite case inside an existing test?

robrap commented:
Done with pass.  Thanks @andy-armstrong.

andy-armstrong commented:
@robrap @dianakhuang @catong @marcotuts thanks for all the great feedback. I've addressed your comments and have kicked-off a sandbox update. For the moment I've gone with @robrap's suggested name for the setting ("Allow Proctoring Opt-Out"), and I've updated the help text to match. We can do final tweaks to these strings on Monday. Please re-review when you have the time.

andy-armstrong commented on edx_proctoring/templates/proctored_exam/proctoring_opt_out_button.html:
```diff
@@ -0,0 +1,13 @@
+{% load i18n %}
+
+{% if allow_proctoring_opt_out %}
```
@robrap I decided to include the test both here and in the parent templates. It seemed risky not to have it here, but it seemed unclear in the parent templates not to show that the include was conditional. Let me know if you'd rather I only do the test in one place or the other.

robrap replied:
Your reasoning makes sense to me.

robrap commented:
👍  Once you resolve the final language question with @catong regarding [the footer commented on here](https://github.com/edx/edx-proctoring/pull/315#discussion_r85747122).

robrap commented:
Oh, and also updating the `secondary-action` class to match the other PR. :)

andy-armstrong commented:
@robrap I've addressed the final issues and squashed my commits. 

@dianakhuang this is ready for you to review now. I've updated my sandbox to have the latest version.

dianakhuang commented:
👍  from me.
